Optimiser.fs

```diff
@@ -180,40 +182,60 @@ module Utils =
     /// <returns>
     ///     A sequence of optimisers to run.
     /// </returns>
-    let mkOptimiserSet opts removes adds =
+    let mkOptimiserSet all_opts opts =
         let config = config()
-        if Set.contains "all" removes
-        then
-             if config.verbose
-             then eprintfn "note: all optimisations disabled"
-
-             []
-        else if Set.contains "all" adds
-        then
-            if config.verbose
-            then eprintfn "note: all optimisations enabled"
-
-            List.map (fun (_, _, ofun) -> ofun) opts
-        else
-            opts
-            |> Seq.choose
-                   (fun (name, on, f) ->
-                        let on' = (on || optAllowed adds name)
-                                  && (not (optAllowed removes name))
-
-                        if (config.verbose && (on' <> on))
-                        then eprintfn "note: optimisation %s forced %s"
-                                      name
-                                      (if on' then "on" else "off")
-
-                        if on' then Some f else None)
-            (* We use List instead of Seq to make sure the above evaluates
-               only once. *)
-            |> List.ofSeq
+        let optimisationSet = new HashSet<string>();
+        // try add or remove from prefix
+        let addFromPrefix prefix =
+            for (optName : string, _, _) in all_opts do
+                if optName.StartsWith(prefix) then
+                    if config.verbose && not (optimisationSet.Contains(optName)) then
+                        eprintfn "note: forced %s on" optName
+
+                    ignore <| optimisationSet.Add(prefix)
+
+        let removeFromPrefix prefix =
+            for (optName, _, _) in all_opts do
+                if optName.StartsWith(prefix) then
+                    if config.verbose && optimisationSet.Contains(optName) then
+                        eprintfn "note: forced %s off" optName
+
+                    ignore <| optimisationSet.Remove(prefix)
+
+        for (optName, enabledByDefault, _) in all_opts do
+            if enabledByDefault then
+                ignore <| optimisationSet.Add(optName)
```

Comment:
Could be `List.iter (fun (optName, enabledByDefault, _) -> if EnabledByDefault then optimisationSet.Add(optName))`, I think, but it doesn't really matter.